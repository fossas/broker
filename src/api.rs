//! Interaction with remote APIs live here.
//! See each submodule for details.

pub mod code_host;
pub mod fossa;
pub mod http;
pub mod ssh;
