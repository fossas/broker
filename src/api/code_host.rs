//! This module provides functionality for integrating with external code hosts.

use std::time::Duration;

use derive_more::{AsRef, Display, From};
use derive_new::new;
use error_stack::{IntoReport, Report, ResultExt};
use getset::{CopyGetters, Getters};
use url::Url;

use crate::ext::error_stack::{DescribeContext, ErrorHelper};

pub mod git;

/// Errors that are possibly surfaced during validation of config values.
#[derive(Debug, thiserror::Error)]
pub enum ValidationError {
    /// Poll interval is parsed from a user-provided string.
    #[error("validate poll interval")]
    PollInterval,

    /// The provided URL is not valid.
    #[error("validate endpoint URL")]
    ValidateEndpoint,
}

/// Validated config values for external code host integrations.
#[derive(Debug, Clone, PartialEq, Eq, AsRef, From, new)]
pub struct Config(Vec<Integration>);

/// Validated endpoint for a code host.
#[derive(Debug, Clone, PartialEq, Eq, AsRef, From, Display, new)]
pub struct Endpoint(Url);

impl TryFrom<String> for Endpoint {
    type Error = Report<ValidationError>;

    fn try_from(input: String) -> Result<Self, Self::Error> {
        Url::parse(&input)
            .into_report()
            .describe_lazy(|| format!("provided input: '{input}'"))
            .help("the url provided must be absolute and must contain the protocol")
            .change_context(ValidationError::ValidateEndpoint)
            .map(Endpoint)
    }
}

/// Each integration has metadata (for example, how often it should be polled)
/// along with its protocol (which describes how to download the code so it can be analyzed).
///
/// This type stores this combination of data.
#[derive(Debug, Clone, PartialEq, Eq, Getters, CopyGetters, new)]
pub struct Integration {
    /// The interval at which Broker should poll the remote code host for whether the code has changed.
    #[getset(get_copy = "pub")]
    poll_interval: PollInterval,

    /// The protocol Broker uses to communicate with the remote code host.
    #[getset(get = "pub")]
    protocol: Protocol,
}

/// Code is stored in many kinds of locations, from git repos to
/// random FTP sites to DevOps hosts like GitHub.
///
/// To handle this variety, Broker uses a predefined list
/// of supported protocols (this type),
/// which are specialized with configuration unique to those integrations.
#[derive(Debug, Clone, PartialEq, Eq, From, new)]
pub enum Protocol {
    /// Integration with a code host using the git protocol.
    Git(git::Transport),
}

/// Specifies the maximum age for an observability artifact.
#[derive(Debug, Copy, Clone, PartialEq, Eq, AsRef, From, new)]
pub struct PollInterval(Duration);

impl TryFrom<String> for PollInterval {
    type Error = Report<ValidationError>;

    fn try_from(value: String) -> Result<Self, Self::Error> {
        parse_duration::parse(&value)
            .into_report()
            .change_context(ValidationError::PollInterval)
            .describe_lazy(|| format!("provided value: {value}"))
            .map(PollInterval)
    }
}
