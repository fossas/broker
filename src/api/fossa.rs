//! Interactions and data types for the FOSSA API live here.

use derive_more::{AsRef, Display, From};
use derive_new::new;
use error_stack::{IntoReport, Report, ResultExt};
use getset::Getters;
use url::Url;

use crate::ext::{
    error_stack::{DescribeContext, ErrorHelper},
    secrecy::ComparableSecretString,
};

/// Errors that are possibly surfaced during validation of config values.
#[derive(Debug, thiserror::Error)]
pub enum ValidationError {
    /// The provided URL is not valid.
    #[error("validate endpoint URL")]
    Endpoint,

    /// The provided API key is not valid.
    #[error("validate API key")]
    ApiKey,

    /// The value provided to parse is empty.
    #[error("provided value is empty")]
    ValueEmpty,
}

/// Validated config values for the FOSSA API.
#[derive(Debug, Clone, PartialEq, Eq, Getters, new)]
#[getset(get = "pub")]
pub struct Config {
    /// The endpoint for the FOSSA backend.
    endpoint: Endpoint,

    /// The key used when interacting with the FOSSA backend.
    key: Key,
}

/// The URL to the FOSSA endpoint.
#[derive(Debug, Clone, PartialEq, Eq, AsRef, Display, From, new)]
pub struct Endpoint(Url);

impl TryFrom<String> for Endpoint {
    type Error = Report<ValidationError>;

    fn try_from(input: String) -> Result<Self, Self::Error> {
        Url::parse(&input)
            .into_report()
            .describe_lazy(|| format!("provided input: '{input}'"))
            .help("the url provided must be absolute and must contain the protocol, for example 'https://app.fossa.com'")
            .change_context(ValidationError::Endpoint)
            .map(Endpoint)
    }
}

/// The FOSSA API key.
#[derive(Debug, Clone, PartialEq, Eq, AsRef, From, new)]
pub struct Key(ComparableSecretString);

impl TryFrom<String> for Key {
    type Error = Report<ValidationError>;

    fn try_from(input: String) -> Result<Self, Self::Error> {
        if input.is_empty() {
            Err(Report::new(ValidationError::ValueEmpty))
                .describe_lazy(|| format!("provided input: '{input}'"))
                .help("use an API key from FOSSA here: https://app.fossa.com/account/settings/integrations/api_tokens")
                .change_context(ValidationError::ApiKey)
        } else {
            let secret = ComparableSecretString::from(input);
            Ok(Key(secret))
        }
    }
}
