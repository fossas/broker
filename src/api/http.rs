//! Interact with remote services over HTTP!

use std::fmt::Display;

use derive_more::From;
use derive_new::new;
use serde::{Deserialize, Serialize};

use crate::ext::secrecy::ComparableSecretString;

/// HTTP authentication can be performed either with a header or via 'HTTP Basic'.
#[derive(Debug, Clone, PartialEq, Eq, From, Deserialize, Serialize, new)]
pub enum Auth {
    /// Uses a header value for authentication.
    Header(ComparableSecretString),

    /// Uses HTTP Basic to perform authentication.
    Basic {
        /// The username for authentication.
        username: String,

        /// The password for authentication.
        password: ComparableSecretString,
    },
}

impl Display for Auth {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Auth::Header(header) => write!(f, "authorization header {header}"),
            Auth::Basic { username, password } => {
                write!(f, "username '{username}' and password '{password}'")
            }
        }
    }
}
