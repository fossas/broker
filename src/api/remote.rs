//! This module provides functionality for integrating with "remotes", which are external code hosts.
//!
//! # "Remote" as a concept
//!
//! An example of a [`Remote`] is Github; it's a place where code is stored remotely.
//! [`Remote`] as a concept is abstracted over protocol implementation:
//! for example, a code host accessed via `ssh` or via `http` are both simply a [`Remote`].
//!
//! Their connection protocol (and any specifics, like authentication) is then specified via
//! [`Protocol`], which is usually wrapped inside an [`Integration`], forming the primary interaction
//! point for this module.

use std::time::Duration;

use derive_more::{AsRef, Display, From};
use derive_new::new;
use error_stack::{report, IntoReport, Report, ResultExt};
use getset::{CopyGetters, Getters};
use humantime::parse_duration;
use url;

use crate::ext::error_stack::{DescribeContext, ErrorHelper};

pub mod git;

/// Errors that are possibly surfaced during validation of config values.
#[derive(Debug, thiserror::Error)]
pub enum ValidationError {
    /// Poll interval is parsed from a user-provided string.
    #[error("validate poll interval")]
    PollInterval,

    /// The provided remote is not valid.
    #[error("validate remote location")]
    Remote,

    /// The provided value is empty.
    #[error("value is empty")]
    ValueEmpty,
}

/// Validated config values for external code host integrations.
#[derive(Debug, Default, Clone, PartialEq, Eq, AsRef, From, new)]
pub struct Config(Vec<Integration>);

/// Validated remote location for a code host.
#[derive(Debug, Clone, PartialEq, Eq, AsRef, Display, new)]
pub struct Remote(String);

impl Remote {
    /// parses a Remote and returns a Result<Url>
    pub fn parse(&self) -> Result<url::Url, url::ParseError> {
        url::Url::parse(&self.0)
    }
}

impl TryFrom<String> for Remote {
    type Error = Report<ValidationError>;

    fn try_from(input: String) -> Result<Self, Self::Error> {
        // Different remotes have different semantics-
        // we can't guarantee this is actually a well formatted URL.
        // Just validate that it's not empty.
        if input.is_empty() {
            Err(report!(ValidationError::ValueEmpty))
                .describe_lazy(|| format!("provided input: '{input}'"))
        } else {
            Ok(Remote(input))
        }
        .help("the remote location may not be empty")
        .change_context(ValidationError::Remote)
    }
}

/// Each integration has metadata (for example, how often it should be polled)
/// along with its protocol (which describes how to download the code so it can be analyzed).
///
/// This type stores this combination of data.
#[derive(Debug, Clone, PartialEq, Eq, Getters, CopyGetters, new)]
pub struct Integration {
    /// The interval at which Broker should poll the remote code host for whether the code has changed.
    #[getset(get_copy = "pub")]
    poll_interval: PollInterval,

    /// The protocol Broker uses to communicate with the remote code host.
    #[getset(get = "pub")]
    protocol: Protocol,
}

/// Code is stored in many kinds of locations, from git repos to
/// random FTP sites to DevOps hosts like GitHub.
///
/// To handle this variety, Broker uses a predefined list
/// of supported protocols (this type),
/// which are specialized with configuration unique to those integrations.
#[derive(Debug, Clone, PartialEq, Eq, From, new)]
pub enum Protocol {
    /// Integration with a code host using the git protocol.
    Git(git::Transport),
}

/// Specifies the maximum age for an observability artifact.
#[derive(Debug, Copy, Clone, PartialEq, Eq, AsRef, From, new)]
pub struct PollInterval(Duration);

impl TryFrom<String> for PollInterval {
    type Error = Report<ValidationError>;

    fn try_from(value: String) -> Result<Self, Self::Error> {
        parse_duration(&value)
            .into_report()
            .change_context(ValidationError::PollInterval)
            .describe_lazy(|| format!("provided value: {value}"))
            .map(PollInterval)
    }
}
