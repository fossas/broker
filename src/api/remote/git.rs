pub mod repository;
pub mod transport;
use std::fmt::Display;

use derive_new::new;
use serde::{Deserialize, Serialize};

/// A git reference's type (branch or tag)
#[derive(Debug, Clone, Hash, Eq, PartialEq, Serialize, Deserialize, new)]
pub enum Reference {
    /// A branch
    Branch {
        /// The name of the branch
        name: String,

        /// The head commit of the branch
        head: String,
    },

    /// A tag
    Tag {
        /// The name of the tag
        name: String,
        /// The commit that the tag points at
        commit: String,
    },
}

impl Display for Reference {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Reference::Branch { name, head } => {
                write!(f, "branch::{name}@{head}")
            }
            Reference::Tag { name, commit } => {
                write!(f, "tag::{name}@{commit}")
            }
        }
    }
}

impl Reference {
    /// Retrieves name of reference
    pub fn name(&self) -> &String {
        match self {
            Self::Branch { name, .. } => name,
            Self::Tag { name, .. } => name,
        }
    }

    /// Generate a canonical state for the reference.
    pub fn as_state(&self) -> &[u8] {
        match self {
            Reference::Branch { head, .. } => head.as_bytes(),
            Reference::Tag { commit, .. } => commit.as_bytes(),
        }
    }

    /// Generate a representation for the reference suitable for use when
    /// creating database coordinates.
    pub fn for_coordinate(&self) -> String {
        match self {
            Reference::Branch { name, head } => format!("branch:{name}@{head}"),
            Reference::Tag { name, commit } => format!("tag:{name}@{commit}"),
        }
    }
}
