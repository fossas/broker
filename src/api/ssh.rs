//! Interact with remote services over SSH.

use std::{fmt::Display, path::PathBuf};

use derive_more::From;
use derive_new::new;
use serde::{Deserialize, Serialize};

use crate::ext::secrecy::ComparableSecretString;

/// SSH authentication can be performed either with a key file or with a static key.
#[derive(Debug, Clone, PartialEq, Eq, From, Deserialize, Serialize, new)]
pub enum Auth {
    /// Uses a key file on disk to perform authentication.
    KeyFile(PathBuf),

    /// Uses a value stored in the config file (and therefore in memory) to perform authentication.
    KeyValue(ComparableSecretString),
}

impl Display for Auth {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Auth::KeyFile(path) => write!(f, "key file at '{}'", path.display()),
            Auth::KeyValue(key) => write!(f, "private key {key}"),
        }
    }
}
