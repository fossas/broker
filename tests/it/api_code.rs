//! Tests for `api::code` functionality.

use broker::api::remote::{PollInterval, Remote};
use proptest::{prop_assert, prop_assert_eq};
use test_strategy::proptest;

use crate::{assert_error_stack_snapshot, helper::duration::DurationInput};

#[proptest]
fn validate_poll_interval(input: DurationInput) {
    let user_input = input.to_string();
    match PollInterval::try_from(user_input.clone()) {
        Ok(validated) => prop_assert_eq!(
            validated.as_ref(),
            &input.expected_duration(),
            "tested input: {:?}",
            input
        ),
        Err(err) => prop_assert!(
            false,
            "unexpected parsing error '{err:#}' for input '{user_input}'"
        ),
    }
}

#[test]
fn validate_poll_interval_empty() {
    let input = String::from("");
    assert_error_stack_snapshot!(
        &input,
        PollInterval::try_from(input).expect_err("must have failed validation")
    )
}

#[proptest]
fn validate_remote(#[strategy("\\PC+")] input: String) {
    match Remote::try_from(input.clone()) {
        Ok(validated) => prop_assert_eq!(validated.as_ref(), &input, "tested input: {:?}", input),
        Err(err) => prop_assert!(
            false,
            "unexpected parsing error '{err:#}' for input '{input}'"
        ),
    }
}

#[test]
fn validate_remote_empty() {
    let input = String::from("");
    assert_error_stack_snapshot!(
        &input,
        Remote::try_from(input).expect_err("must have failed validation")
    )
}
