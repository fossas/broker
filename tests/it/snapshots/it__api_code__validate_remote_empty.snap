---
source: tests/it/api_code.rs
expression: "Remote::try_from(input).expect_err(\"must have failed validation\")"
info: ""
---
validate remote location
├╴at {source location}
│
╰─▶ value is empty
    ├╴at {source location}
    ├╴context: provided input: ''
    ╰╴help: the remote location may not be empty
